Sources/PointFree/Account/Invite.swift

```diff
@@ -0,0 +1,208 @@
+import Css
+import Either
+import Foundation
+import Html
+import HtmlCssSupport
+import HttpPipeline
+import HttpPipelineHtmlSupport
+import Optics
+import Prelude
+import Styleguide
+import Tuple
+
+let showInviteMiddleware =
+  // TODO: need to validate that current user doesnt already have a subscription
+  // TODO: validate that current user is not inviter
+  requireTeamInvite
+    <| currentUserMiddleware
+    >-> writeStatus(.ok)
+    >-> respond(showInviteView.contramap(lower))
+
+let revokeInviteMiddleware =
+  requireTeamInvite
+    <<< requireUser
+    <| { conn in
+      // TODO: validate that current user owns team invite
+      AppEnvironment.current.database.deleteTeamInvite(get2(conn.data).id)
+        .run
+        .flatMap(const(conn |> redirect(to: path(to: .team(.show)))))
+}
+
+let resendInviteMiddleware =
+  requireTeamInvite
+    <<< requireUser
+    <| { conn in
+      parallel(sendInviteEmail(invite: get2(conn.data), inviter: get1(conn.data)).run)
+        .run({ _ in })
+      return conn |> redirect(to: path(to: .team(.show)))
+}
+
+let acceptInviteMiddleware =
+  requireTeamInvite
+    <<< requireUser
+    <| { conn in
+      let (currentUser, teamInvite) = lower(conn.data)
+
+      // VERIFY: need to validate that current user doesnt already have a subscription
+      let invitee = pure(currentUser)
+        .flatMap(validateDoesNotHaveSubscription)
+
+      // VERIFY: need to validate that current user is not the inviter
+      let inviter = AppEnvironment.current.database
+        .fetchUserById(teamInvite.inviterUserId)
+        .mapExcept(requireSome)
+        .flatMap(validateIsNot(currentUser: currentUser))
+
+      let sendInviterEmailOfAcceptance = parallel(
+        inviter
+          .run
+          .flatMap { errorOrInviter in
+            errorOrInviter.right.map { inviter in
+              sendEmail(
+                to: [inviter.email],
+                subject: "\(currentUser.name) has accepted your Point-Free team invitation!",
+                content: inj2(inviteeAcceptedEmailView.view((inviter: inviter, invitee: currentUser)))
+                )
+                .run
+                .map(const(unit))
+              }
+              ?? pure(unit)
+      })
+
+      // VERIFY: user is subscribed
+      let subscription = inviter
+        .map(^\.subscriptionId)
+        .mapExcept(requireSome)
+        .flatMap { AppEnvironment.current.database.fetchSubscriptionById($0) }
+        .mapExcept(requireSome)
+        .flatMap { subscription in
+          AppEnvironment.current.database.addUserIdToSubscriptionId(currentUser.id, subscription.id)
+      }
+
+      // VERIFY: only do this if the invite was successfully taken
+      let deleteInvite = parallel(
+        subscription
+          // TODO: should `const` be @autoclosure so that we can do:
```

Comment:
I think we probably need a way of doing both strict and lazy evaluation. Scalaz 8 IO does this by having both `IO.now` (strict) and `IO.point` (lazy). The applicative `pure` is actually `now` (for performance, I think), so I bet `const` would want to strict, too, with how often we use it.